src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -74,13 +74,17 @@ typesOf ::
   [(Ident, Expr)] ->
   m [(Ident, (Expr, Type))]
 typesOf bindingGroupType moduleName vals = do
-  tys <- fmap tidyUp . liftUnifyWarnings replace $ do
+  ((tys, warnings), subst) <- withEmptySubstitution . reflectErrors . censor (const mempty) . reifyErrors . listen $ do
     (untyped, typed, dict, untypedDict) <- typeDictionaryForBindingGroup moduleName vals
     ds1 <- parU typed $ \e -> checkTypedBindingGroupElement moduleName e dict
     ds2 <- forM untyped $ \e -> typeForBindingGroupElement e dict untypedDict
     return (map (\x -> (False, x)) ds1 ++ map (\x -> (True, x)) ds2)
 
-  forM tys $ \(shouldGeneralize, (ident, (val, ty))) -> do
+  let (holeErrors, warnings') = partition isHoleError $ map (replace subst) $ runMultipleErrors warnings
```

Comment:
I think I'd be fine with this way of treating warnings if we could just pull it out into a combinator. Something like 

```haskell
escalate 
  :: (MonadWriter [MultipleErrors] m, MonadError [MultipleErrors] m)
  => (ErrorMessage -> Bool)
  -> m a
  -> m a
```